src/GraphQL/Execution/SubscriptionExecutionStrategy.cs

```diff
@@ -0,0 +1,318 @@
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance with the specified execution strategy for child nodes.
+        /// </summary>
+        public SubscriptionExecutionStrategy(IExecutionStrategy baseExecutionStrategy)
+        {
+            _baseExecutionStrategy = baseExecutionStrategy ?? throw new ArgumentNullException(nameof(baseExecutionStrategy));
+        }
+
+        /// <summary>
+        /// Gets a static instance of <see cref="SubscriptionExecutionStrategy"/>.
+        /// </summary>
+        public static SubscriptionExecutionStrategy Instance { get; } = new();
+
+        /// <summary>
+        /// Executes a GraphQL subscription request and returns the result. The result consists
+        /// of one or more streams of GraphQL responses, returned within <see cref="SubscriptionExecutionResult.Streams"/>.
+        /// No serializable <see cref="ExecutionResult"/> is directly returned.
+        /// <br/><br/>
+        /// Keep in mind that if a scoped context is passed into <see cref="ExecutionContext.RequestServices"/>,
+        /// and if it is disposed after the initial execution, node executions of subsequent data events will contain
+        /// the disposed <see cref="ExecutionContext.RequestServices"/> instance and hence be unusable.
+        /// <br/><br/>
+        /// If scoped services are needed, it is recommended to override
+        /// <see cref="ProcessDataAsync(ExecutionContext, ExecutionNode, object?)"/>
+        /// with an implementation that creates a service scope during execution of the data events.
+        /// </summary>
+        public override async Task<ExecutionResult> ExecuteAsync(ExecutionContext context)
+        {
+            var rootType = GetOperationRootType(context);
+            var rootNode = BuildExecutionRootNode(context, rootType);
+
+            var streams = await ExecuteSubscriptionNodesAsync(context, rootNode.SubFields!).ConfigureAwait(false);
+
+            // if execution is successful, errors and extensions are not returned per the graphql-ws protocol
+            // if execution is unsuccessful, the DocumentExecuter will add context errors to the result
+
+            return new SubscriptionExecutionResult(context)
+            {
+                Executed = true,
+                Streams = streams,
+            };
+        }
+
+        private async Task<IDictionary<string, IObservable<ExecutionResult>>?> ExecuteSubscriptionNodesAsync(ExecutionContext context, ExecutionNode[] nodes)
+        {
+            var streams = new Dictionary<string, IObservable<ExecutionResult>>();
+
+            foreach (var node in nodes)
+            {
+                var stream = await ResolveEventStreamAsync(context, node).ConfigureAwait(false);
+                if (stream == null)
+                    return null;
```

Comment:
🤔  Why so?

Before:
```
 if (stream != null)
                    streams[node.Name!] = stream;
```

Reply:
One more concern - returned observables _may_ implement `IDisposable` and _may_ be "hot" observables so silently discarding streams dictionary may have performance penalty (both CPU and memory).

Reply:
First, it is against GraphQL spec to ever request more than one subscription in a single request.  I had previously changed the code to make it impossible to resolve multiple streams at once, but then reverted the design, as there's really no reason why the execution strategy cares one way or the other.

Second, the overall design of GraphQL is such that requesting an object will always return a JSON response that matches said object.  So `{ hero villan }` is always going to return `{ "hero": (...), "villan": (...) }`.  Only in the case of a nullable field should it be able to return `{ "hero": (...), "villan": null }` in the event of an error.  Never `{ "hero": (...) }` without `villan` in the map.  Rather, the `null` bubbles.

So the act of `if (stream != null)` is clearly wrong, in keeping with this philosophy.  My `return null` behavior acts as though the fields were non-null, which is more convenient for the server.  The other option is to always set it in the dictionary and remove the `if` condition.  For the server to meet spec, and even for it to work at all, not only should it ensure that only a single field has executed (we already have a validation rule for that), but it would also need to check the dictionary to be sure non-null `IObservable` instances were returned.

I'm happy to make the change if you think that would be more appropriate.  But I believe the prior behavior is wrong.

Reply:
As for an `IObservable` being a 'hot' object, while this is possible, this is clearly not the design of `IObservable` and there are no existing provisions to check for it being a disposable object.  Not only that, it is very likely that it is disposable, and it is very important that it not be disposed.  For example, assume you set up observables in your application like this:

```cs
// startup.cs
var saleBroadcaster = new Subject<Sale>();
services.AddSingleton<IObserver<Sale>>(saleBroadcaster); //use this to send notifications
services.AddSingleton<IObservable<Sale>>(saleBroadcaster); //use this to subscribe
```

With those few lines of code, any object can pull `IObservable<Sale>` and subscribe (in GraphQL subscriptions or perhaps an `IHostedService` implementation), or send notifications through `IObserver<Sale>`.  But if they were to attempt to dispose of it, it would shut down subscriptions for the entire application.  Clearly, there is no intent that `IObservable<T>` should be cast to `IDisposable` and disposed of after use, and in my opinion, it would likely break many (if not most) implementations.

Reply:
I'm OK returning `null`. What about discarded observables?

Reply:
OK. I think we have here the same story as we had about `IAsyncEnumerable` returned from resolvers. We don't own objects returned.